Supersede RepairableNear with ProximityLinkHost

PunkPun opened the pull request:
part of #20168

 Add remote hosts and replace `RepairableNear` for RA naval. Supersedes #20169

ProximityLinkHost has 2 modes controlled by `OccupyClient`. It determines wether docking is carried out via activities on the client, or via tick method on the host

PunkPun committed:
Add ProximityLinkHost

penev92 commented:
There were discussions about "remote docking" not being a thing we want.

PunkPun commented:
Remote docking is needed for various things. Something like repairing could be handled with a fake RemoteDock and GrantCondition in range with that condition enabling self heal. Or repair could be done with one of the 4 methods this PR introduces. Fwiw, this should already allow for fake docks.

Next are things that actually require a connection. Something like depositing resources, paying for ammo, paying for repair, paying for upgrades. Or, for example when only a limited number of vehicles can be repaired (because there are not enough repair drones).

Another example is when you want an animation to play on a structure when something is repairing from it nearby.

abcdefg30 commented on OpenRA.Mods.Common/Activities/RemoteDockSequence.cs:
```diff
@@ -0,0 +1,64 @@
+
+		protected override void OnFirstRun(Actor self)
+		{
+			if (!IsCanceling && !DockHostActor.IsDead && DockHostActor.IsInWorld
+				&& (self.CenterPosition - DockHost.DockPosition).Length <= CloseEnough
```
This should be extracted to a helper function instead of being written twice. That then also helps readability.

abcdefg30 replied:
This is still not resolved.

abcdefg30 commented on OpenRA.Mods.Common/Activities/RemoteDockSequence.cs:
```diff
@@ -0,0 +1,64 @@
+			else
+			{
+				DockClient.UnreserveHost();
+			}
```
```suggestion
				DockClient.UnreserveHost();
```

abcdefg30 commented on OpenRA.Mods.Common/Activities/RemoteDockSequence.cs:
```diff
@@ -0,0 +1,64 @@
+				&& (self.CenterPosition - DockHost.DockPosition).Length <= CloseEnough
+				&& !DockClient.OnDockTick(self, DockHostActor, DockHost))
+			{
+				return false;
```
Shouldn't this also unreserve the host if any of the conditions except `OnDockTick` are not fulfilled?

abcdefg30 replied:
Ideally, this should happen in `OnLastRun` of the activity instead of manually handling the all cases except success.

PunkPun replied:
unreserve is called inside `OnDockCompleted`, perhaps that should be documented

abcdefg30 replied:
Then it can always run in `OnLastRun` regardless of success or not. (Unreserving twice shouldn't matter.)

PunkPun replied:
I think the current implementation is cleaner

abcdefg30 replied:
With the new helper method it indeed looks clean enough. That opens one more question though: Should it also call OnDock*Completed* when docking fails due to cancelling? (I assume you made it this way because cancelling also counts as being completed?)

PunkPun replied:
The logic is: if docking was started, it must be completed. It's only not fully true in cases of actor death as it doesn't really matter for the dead actor.

Canceling would be considered premature completion.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Buildings/RemoteDockHost.cs:
```diff
@@ -0,0 +1,341 @@
+		[Desc("Does the client need to be preoccupied with docking?")]
+		public readonly bool OccupyClient = false;
+
+		[Desc("Automatically dock clients in " + nameof(Range) + ".")]
```
What is the use-case for this? This should at least be a subclass of this trait to reduce complexity in the implementation.

abcdefg30 replied:
That will likely also be better for performance as most use cases (i.e. non-auto-dock) won't need all the interface calls.

PunkPun replied:
Autodock could be used like a heal area. For instance the dock could have a repair drone, and when you move within range of the dock, the drone moves over and heals you.

I'm not sure how we want to implement autodock so I just went with 4 types of remote docking in 2, I can see all 4 being useful. Though, I would suspect that traits inheriting this would likely only want 1 method

abcdefg30 replied:
RemoteDockHost should just have the basic common functionality. A new AutoRemoteDockHost or RemoteAutoDockHost trait has to be created for all the special bits that involve auto-docking.

PunkPun replied:
pushed into a RemoteAutoDockHost trait

abcdefg30 commented on OpenRA.Mods.Common/Traits/Buildings/RemoteDockHost.cs:
```diff
@@ -0,0 +1,341 @@
+					self.World.ActorMap.UpdateProximityTrigger(proximityTrigger, cachedPosition, cachedRange, WDist.Zero);
+				}
+
+				// Activelly search for nearby clients.
```
```suggestion
				// Actively search for nearby clients.
```

abcdefg30 commented on OpenRA.Mods.Common/Traits/Buildings/RemoteDockHost.cs:
```diff
@@ -0,0 +1,341 @@
+
+		protected readonly List<DockClientManager> ReservedDockClients = new();
+		protected readonly List<TraitPair<DockClientManager>> ProximityClients = new();
+		protected readonly List<(TraitPair<DockClientManager>, long)> WaitingClients = new();
```
The tuple parameters should be named.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Buildings/RemoteDockHost.cs:
```diff
@@ -0,0 +1,341 @@
+
+		void ActorExited(Actor a)
+		{
+			var client = ProximityClients.FirstOrDefault(p => p.Actor == a);
```
```suggestion
			var client = ProximityClients.Find(p => p.Actor == a);
```

abcdefg30 commented on OpenRA.Mods.Common/Traits/Buildings/RemoteDockHost.cs:
```diff
@@ -0,0 +1,341 @@
+			}
+			else
+			{
+				// When auto-docking, this function be repeadedly called.
```
```suggestion
				// When auto-docking, this function will be called repeatedly.
```
?

Also this really shouldn't be handled by using `QueueDockActivity` but using another method, since we are not queuing a docking activity. (That or the method name needs to be adjusted.)

PunkPun replied:
This is an interface implementation, so I'm not sure wether this should be changed

abcdefg30 commented on OpenRA.Mods.Common/Traits/World/ActorMap.cs:
```diff
@@ -140,7 +140,7 @@ public void Tick(ActorMap am)
 				currentActors.UnionWith(
 					am.ActorsInBox(position - delta, position + delta)
-					.Where(a => (a.CenterPosition - position).HorizontalLengthSquared < range.LengthSquared
+					.Where(a => (a.CenterPosition - position).HorizontalLengthSquared <= range.LengthSquared
```
Leaving this comment for other reviewers: This will need some consideration as well. (As it impacts more than just docking.)

penev92 commented:
I remember 3 people speaking against "remote docking" in Discord.
Just mentioning here so there's no "there weren't any concerns on the PR".

abcdefg30 commented on OpenRA.Mods.Common/Traits/Buildings/RemoteDockHost.cs:
```diff
@@ -0,0 +1,226 @@
+		protected virtual void Tick(Actor self)
+		{
+			// We are not allowed to use activities.
+			if (!Info.OccupyClient)
```
Please transform this to
```
			if (Info.OccupyClient)
				return;
```

to reduce nesting.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Buildings/RemoteDockHost.cs:
```diff
@@ -0,0 +1,226 @@
+
+					if (WaitingClients[i].Time < self.World.WorldTick)
+					{
+						if (clientActor != null && !clientActor.IsDead && clientActor.IsInWorld && !IsTraitDisabled && !preventDock
+							&& (clientActor.CenterPosition - DockPosition).HorizontalLengthSquared <= Info.Range.LengthSquared
```
This would also benefit from a helper method. Move everything except the `preventDock` check into it and then re-use it below as well.

PunkPun commented:
- Removed `RemoteAutoDockHost` and renamed `RemoteDockHost` -> `ProximityHost`
- I've superseded and removed RepairableNear, so RA naval will now be using a `ProximityHost`
- And also added RequireForceMoveCondition to DockClientManager which then revealed a bug which I fixed here as well

abcdefg30 commented on OpenRA.Mods.Common/Traits/DockOnRepairCursor.cs:
```diff
@@ -0,0 +1,49 @@
+		public override object Create(ActorInitializer init) { return new DockOnRepairCursor(init.Self, this); }
+	}
+
+	public class DockOnRepairCursor
```
Why do we have an extra trait for this? This should be part of the dock client?

PunkPun replied:
Why should it? I can imagine plenty scenarios where you would want this trait but not Repairable. Let's say in a case where healing is granted via condition

PunkPun replied:
It's also not the job of Repairable to give orders, but of DockClientManager

abcdefg30 replied:
Why would healing granted by a condition be covered by this trait? The naming clearly suggests it is about docking. It does not make sense to separate this trait.

PunkPun replied:
> Why would healing granted by a condition be covered by this trait?

the actor that provides the healing aura could have a fake dock. It's one way to make repair cursor work with an alternative healing system

PunkPun committed:
Testcase

PunkPun commented:
We now linkin' instead 🚗

abcdefg30 commented on OpenRA.Mods.Common/Traits/LinkClientBase.cs:
```diff
@@ -0,0 +1,54 @@
+			: base(info)
+		{
+			this.self = self;
+			LinkClientManager = self.TraitOrDefault<LinkClientManager>();
```
```suggestion
			LinkClientManager = self.Trait<LinkClientManager>();
```

since we require it?

PunkPun replied:
it was made optional in 8e7fa267092594ef6e1d5255a5d0f2722f249dbb

abcdefg30 commented on OpenRA.Mods.Common/Traits/LinkClientBase.cs:
```diff
@@ -0,0 +1,54 @@
+	{
+		readonly Actor self;
+
+		public abstract BitSet<LinkType> GetLinkType { get; }
```
```suggestion
		public abstract BitSet<LinkType> LinkType { get; }
```

(This is not a method.)

PunkPun replied:
Then the name matches the class name, though I suppose it's not a style error

abcdefg30 commented on OpenRA.Mods.Common/Traits/LinkOnRepairCursor.cs:
```diff
@@ -0,0 +1,49 @@
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Send actor to link to repair cursor.")]
```
Ok I think I understand what this is for now, this is not for providing the repair cursor when sending a unit to an actor to repair, but when hovering over a repairable unit, right? This makes more sense now, but why isn't it simply part of `RepairableLinkClient` without which it wouldn't properly work anyway?

PunkPun replied:
https://github.com/OpenRA/OpenRA/pull/20665#discussion_r1450788616

Porenutak commented:
Is RepairsUnits and Repairable traits also part of this refactor? because they dont work here.

PunkPun commented:
`Repairable` isn't touched. It will be superseded in the future, however RepairsUnits should work. If it didn't naval units wouldn't be repairing

Porenutak commented:
I have vehicles and Air in mind. But if `Repairable `is for future then ok.

pchote commented on OpenRA.Mods.Common/Traits/LinkOnRepairCursor.cs:
```diff
@@ -0,0 +1,49 @@
+		public override object Create(ActorInitializer init) { return new LinkOnRepairCursor(init.Self, this); }
+	}
+
+	public class LinkOnRepairCursor
```
The abstraction here feels off to me. This should really be part of an OrderTargeter?

PunkPun replied:
Because the modder may want to specify which actors this should work on. This abstraction also unties it repair cursor from Repairable trait https://github.com/OpenRA/OpenRA/pull/20665#discussion_r1450788616

pchote commented on OpenRA.Mods.Common/Traits/Cloak.cs:
```diff
@@ -32,7 +32,7 @@ public enum UncloakType
 		Heal = 128,
 		SelfHeal = 256,
-		Dock = 512,
+		Link = 512,
```
Should this be split into `LinkHost` and `LinkClient`?

PunkPun replied:
I'm not sure

PunkPun replied:
It would only matter if an actor would both be able to be linked to and to link itself

pchote commented on OpenRA.Mods.Common/Activities/GenericDockSequence.cs:
```diff
@@ -24,37 +24,37 @@ public class GenericDockSequence : Activity
-		protected readonly IDockClientBody DockClientBody;
+		protected readonly Actor LinkHostActor;
+		protected readonly ILinkHost LinkHost;
+		protected readonly WithDockingOverlay LinkHostSpriteOverlay;
```
Renaming variables from Dock to Link doesn't feel right - this activity (and the `DockHost` trait) are the concrete implementation of the docking behaviour, so it would be clearer if the variable names reflected that.

The way I picture it is that there is the base "Link" concept, of which `ProximityLink` and `DockHost` are concrete implementations. Things that don't care about proximity-vs-dock should be using the general Link terminology, but things that do care should be picking a side so that its immediately clear to the reader what the expectations are.

PunkPun replied:
unnamed

pchote commented on OpenRA.Mods.Common/Activities/FindAndDeliverResources.cs:
```diff
@@ -122,10 +122,10 @@ public override bool Tick(Actor self)
 				if (lastproc != null)
 				{
-					var deliveryLoc = self.World.Map.CellContaining(lastproc.DockPosition);
+					var deliveryLoc = self.World.Map.CellContaining(lastproc.LinkPosition);
```
This is a good example of my comment above: does a `LinkPosition` even make sense for a proximity-style link? This needs to become an implementation detail of the dock style linkage, and then the proximity style linkage would need its own version that considers all suitable cells within range.

PunkPun replied:
it makes sense in that for proximity docking this is the centre of the proximity dock

pchote replied:
The problem is that its at best misleading, and at worst likely to lead to completely wrong behaviour.

Consider a LinkHost that's on the wrong side of an impassable cliff boundary. The host exposes a LinkPosition property, so its completely reasonable to expect that we can/should evaluate a pathfind operation to that position. Since that position isn't pathable, the client won't attempt to Link. This is the correct and expected behaviour for a DockHost. For a ProximityLinkHost it isn't - there are pathable cells within range that the client should be considering instead.

PunkPun replied:
the ProximityLinkHost queues a MoveWithinRange activity so it does consider other positions

pchote replied:
That doesn't address the issue with the abstraction.

PunkPun replied:
I'm not completely happy with the current `ILinkHost` abstraction as well. It restrains to just one possible facing, and a few other things. I think it's a bit out of scope in this PR

pchote commented on OpenRA.Mods.Common/Traits/Buildings/ProximityLinkHost.cs:
```diff
@@ -0,0 +1,252 @@
+				WaitingClients.Add((new TraitPair<LinkClientManager>(clientActor, client), LinkWait + self.World.WorldTick));
+		}
+
+		public virtual void OnLinkCompleted(Actor self, Actor clientActor, LinkClientManager client)
```
IMO the "started" and "completed" terms don't translate very well from Dock to Link (specifically, in this general case, "link completed" may imply that there should also be a "link failed"). Something like `OnLinkCreated` and `OnLinkRemoved` might be clearer?

PunkPun replied:
sure

Mailaender commented on OpenRA.Mods.Common/Traits/LinkClientManager.cs:
```diff
@@ -0,0 +1,372 @@
+	[Desc("Manages LinkClients on the actor.")]
+	public class LinkClientManagerInfo : ConditionalTraitInfo, ILinkClientManagerInfo
+	{
+		[Desc("How long (in ticks) to wait until (re-)checking for a nearby available" + nameof(ILinkHost) + ".")]
```
Not sure if you should mention interfaces in modder documentation.

PunkPun replied:
I think it's better than excluding it

PunkPun committed:
Supersede RepairableNear
